drivers/audio/dmic_nrfx_pdm.c

```diff
@@ -0,0 +1,549 @@
+	uint32_t block_size;
+	struct k_msgq rx_queue;
+	bool request_clock : 1;
+	bool configured    : 1;
+	volatile bool active;
+	volatile bool stopping;
+};
+
+struct dmic_nrfx_pdm_drv_cfg {
+	nrfx_pdm_event_handler_t event_handler;
+	nrfx_pdm_config_t nrfx_def_cfg;
+	enum clock_source {
+		PCLK32M,
+		PCLK32M_HFXO,
+		ACLK
+	} clk_src;
+};
+
+static void free_buffer(struct dmic_nrfx_pdm_drv_data *drv_data, void *buffer)
+{
+	k_mem_slab_free(drv_data->mem_slab, &buffer);
+	LOG_DBG("Freed buffer %p", buffer);
+}
+
+static void event_handler(const struct device *dev, const nrfx_pdm_evt_t *evt)
+{
+	struct dmic_nrfx_pdm_drv_data *drv_data = dev->data;
+	int ret;
+	bool stop = false;
+
+	if (evt->buffer_requested) {
+		void *buffer;
+		nrfx_err_t err;
+
+		ret = k_mem_slab_alloc(drv_data->mem_slab, &buffer, K_NO_WAIT);
+		if (ret < 0) {
+			LOG_ERR("Failed to allocate buffer: %d", ret);
+			stop = true;
+		} else {
+			err = nrfx_pdm_buffer_set(buffer,
+						  drv_data->block_size / 2);
+			if (err != NRFX_SUCCESS) {
+				LOG_ERR("Failed to set buffer: 0x%08x", err);
+				stop = true;
+			}
+		}
+	}
+
+	if (drv_data->stopping) {
+		if (evt->buffer_released) {
+			free_buffer(drv_data, evt->buffer_released);
+		}
+
+		if (drv_data->active) {
+			drv_data->active = false;
+			if (drv_data->request_clock) {
+				(void)onoff_release(drv_data->clk_mgr);
+			}
+		}
+	} else if (evt->buffer_released) {
+		ret = k_msgq_put(&drv_data->rx_queue,
+				 &evt->buffer_released,
+				 K_NO_WAIT);
+		if (ret < 0) {
+			LOG_ERR("No room in RX queue");
+			stop = true;
+
+			free_buffer(drv_data, evt->buffer_released);
+		} else {
+			LOG_DBG("Queued buffer %p", evt->buffer_released);
+		}
+	}
+
+	if (stop) {
+		nrfx_pdm_stop();
+		drv_data->stopping = true;
+	}
+}
+
+static bool is_better(uint32_t freq, uint8_t ratio, uint32_t req_rate,
+		      uint32_t *best_diff, uint32_t *best_freq,
+		      uint32_t *best_rate)
+{
+	uint32_t act_rate = freq / ratio;
+	uint32_t diff = act_rate >= req_rate ? (act_rate - req_rate)
+					     : (req_rate - act_rate);
+
+	LOG_DBG("Freq %u, ratio %u, act_rate %u", freq, ratio, act_rate);
+
+	if (diff < *best_diff) {
+		*best_diff = diff;
+		*best_freq = freq;
+		*best_rate = act_rate;
+		return true;
+	}
+
+	return false;
+}
+
+/* Finds clock settings that give the PCM output rate closest to that requested,
+ * taking into account the hardware limitations.
+ */
+static bool find_suitable_clock(const struct dmic_nrfx_pdm_drv_cfg *drv_cfg,
+				nrfx_pdm_config_t *config,
+				const struct dmic_cfg *pdm_cfg)
+{
+	uint32_t req_rate = pdm_cfg->streams[0].pcm_rate;
+	uint32_t best_freq = 0;
+	uint32_t best_rate = 0;
+	uint32_t best_diff = UINT32_MAX;
+#if NRF_PDM_HAS_RATIO_CONFIG
+	static const struct {
+		uint8_t         ratio_val;
+		nrf_pdm_ratio_t ratio_enum;
+	} ratios[] = {
+		{  64, NRF_PDM_RATIO_64X },
+		{  80, NRF_PDM_RATIO_80X }
+	};
+
+	for (int r = 0; r < ARRAY_SIZE(ratios); ++r) {
+		uint8_t ratio = ratios[r].ratio_val;
+#else
```

Comment:
This is totally unreadable at first glance. Can you move L144 - L233 to a function and call it for number of ratio settings?

Reply:
I refactored that part. I hope it is more reader-friendly now.